ITE drivers/sensor: add tachometer driver for it8xxx2_evb

RuibinChang opened the pull request:
Add tachometer driver for it8xxx2_evb.

Signed-off-by: Ruibin Chang <Ruibin.Chang@ite.com.tw>

RuibinChang committed:
ITE drivers/sensor: add tachometer driver for it8xxx2_evb

Add tachometer driver for it8xxx2_evb.

Signed-off-by: Ruibin Chang <Ruibin.Chang@ite.com.tw>

brockus-zephyr commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:
```diff
@@ -12,11 +12,6 @@
-	aliases {
-		pwm-led0 = &led0;
-		pwm-led1 = &led1;
-	};
```
Does this have anything to do with adding the tach driver?  If not can you please make this a different commit?

RuibinChang replied:
Ok, I will create another commit for cleanup.

brockus-zephyr replied:
thanks

MaureenHelm replied:
These changes still need to be moved to a different commit

RuibinChang replied:
Done

brockus-zephyr commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:
```diff
@@ -27,19 +22,6 @@
-			pwms = <&pwm0 PWM_CHANNEL_0 PWM_POLARITY_NORMAL>;
-			label = "LED1_BLUE";
-		};
-	};
```
Does this have anything to do with adding the tach driver? If not can you please make this a different commit?

RuibinChang replied:
Ok, I will create another commit for cleanup.

brockus-zephyr replied:
thanks

brockus-zephyr commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,282 @@
+
+	switch (tach_ch) {
+	case TACH0_CHANNEL_A:
+		if ((*reg_tswctlr & 0x0C) == IT8XXX2_PWM_T0DVS)
```
Can you create an IT8XXX2_PWM_xxxxx_MASK to remove the 0x0C magic number?

RuibinChang replied:
Done

brockus-zephyr commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,282 @@
+			valid = 1;
+		break;
+	case TACH1_CHANNEL_A:
+		if ((*reg_tswctlr & 0x03) == IT8XXX2_PWM_T1DVS)
```
Can you create an IT8XXX2_PWM_xxxxx_MASK to remove the 0x03 magic number?

RuibinChang replied:
Done

RuibinChang committed:
ITE MAINTAINERS: add tachometer path for compliance check

Add path for drivers/sensor/ite_tach_it8xxx2/*it8xxx2*
and compliance will check it.

Signed-off-by: Ruibin Chang <Ruibin.Chang@ite.com.tw>

MaureenHelm commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,279 @@
+	return valid;
+}
+
+int tach_it8xxx2_sample_fetch(const struct device *dev,
```
This should be `static`

RuibinChang replied:
Done

MaureenHelm commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,279 @@
+
+		if ((tach_ch == TACH0_CHANNEL_A) || (tach_ch == TACH0_CHANNEL_B))
+			/* Clear tachometer0 data valid status */
+			*reg_tswctlr |= IT8XXX2_PWM_T0DVS;
```
Please add braces even for single-line code blocks. See https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

RuibinChang replied:
Done

MaureenHelm commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,279 @@
+			 * - {F1TMRR, F1TLRR} = 0000h denotes Fan Speed is zero
+			 * - The sampling rate (fs) is TACH_FREQ / 128
+			 */
+			val->val1 = (60 * TACH_FREQ / 128 / p / (data->capture));
```
Can `data->capture` ever be zero?

RuibinChang replied:
Yes,  1. when fan stop and tach data is valid, data->capture is zero, or
         2. tach data isn't valid, data->capture is zero.
 
But only data->capture > 0, we transform count to R.P.M by this formula, otherwise we set val->val1 = 0 directly.

RuibinChang commented on boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig:
```diff
@@ -35,3 +35,5 @@ CONFIG_ADC_ITE_IT8XXX2=y
 CONFIG_PWM=y
 CONFIG_PWM_ITE_IT8XXX2=y
+CONFIG_SENSOR=y
+CONFIG_TACH_IT8XXX2=y
```
Ack. I put the conditional in the soc Kconfig.defconfig.series, because our Kconfig.defconfig sources from  Kconfig.defconfig.series.

Ok, I will create aonther PR for the rest.

MaureenHelm replied:
Still need to remove `CONFIG_SENSOR=y` from `boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig`

RuibinChang replied:
We want to enable CONFIG_SENSOR on it8xxx2_evb board, if we remove CONFIG_SENSOR=y in boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig, then where shall I enable it?

RuibinChang replied:
Done

RuibinChang commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,279 @@
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev)	((const struct tach_it8xxx2_config *)(dev)->config)
+#define DRV_DATA(dev)	((struct tach_it8xxx2_data *)(dev)->data)
```
Done

keith-zephyr commented on boards/riscv/it8xxx2_evb/it8xxx2_evb.dts:
```diff
@@ -87,11 +83,24 @@
+&tach0 {
+	status = "okay";
+	/*
+	 * Tach0 module can select input only from [TACH0_CHANNEL_A |
+	 * TACH0_CHANNEL_B]; tach1 module can select input only from
+	 * [TACH1_CHANNEL_A | TACH1_CHANNEL_B]. We don't support that a
+	 * tach dynamically switch input from channel A/B in run time.
```
If you added properties to the device tree to specify the TxDVS and TxCHSEL bits, then you could change the channel property to only track channel A/B.  This way it's not possible for the user to specify cross over the channel inputs.

The sample fetch routine could be simplified to
```
	if (tach_ch_valid(dev, tach_ch)) {
		/* If channel data of tachometer is valid, then save it */
		data->capture = ((*reg_fxtmrr) << 8) | (*reg_fxtlrr);
		/* Clear data valid */
		*reg_tswctlr |= config->dvs_bit;
```
Initialization would look like:
```
	if (tach_ch == TACH_CHANNEL_A) {
		/* Select TACHA output to tachometer */
		*reg_tswctlr &= ~config->chsel_bit;
		/* Clear tachometer data valid status */
		*reg_tswctlr |= config->dvs_bit;
```

The dvs_bit and chsel_bit properties would be set in the it8xxx2.dtsi file.

What do you think?

RuibinChang replied:
Agree and we also needn't the tach_ch_to_pin_idx[] array.
Thanks a lot!

MaureenHelm commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,254 @@
+	switch (tach_ch) {
+	case TACH_CHANNEL_A:
+		if ((*reg_tswctlr & mask) == dvs_bit)
+			valid = 1;
```
add braces here and around line 112

RuibinChang replied:
Done

gmarull replied:
`true`

mbolivar-nordic commented on dts/bindings/tach/ite,it8xxx2-tach.yaml:
```diff
@@ -0,0 +1,39 @@
+    reg:
+        required: true
+
+    dvs_bit:
```
Please use dashes in property names, so `dvs-bit` here...

RuibinChang replied:
Done

mbolivar-nordic commented on dts/bindings/tach/ite,it8xxx2-tach.yaml:
```diff
@@ -0,0 +1,39 @@
+      required: true
+      description: tachometer data valid bit of tswctlr register
+
+    chsel_bit:
```
and `chsel-bit` here.

RuibinChang replied:
Done

mbolivar-nordic commented on dts/bindings/tach/ite,it8xxx2-tach.yaml:
```diff
@@ -0,0 +1,39 @@
+      type: int
+      required: true
+      enum:
+            - 0 #TACH_CHANNEL_A
+            - 1 #TACH_CHANNEL_B
```
Putting the 0=A, 1=B information in the description rather than a comment would be better, so it shows up in https://docs.zephyrproject.org/latest/reference/devicetree/bindings.html

Not a blocking comment

RuibinChang replied:
Done

mbolivar-nordic reviewed (changes requested):
Minor DT changes requested

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,256 @@
+
+/* Device config */
+struct tach_alt_cfg {
+	/* Pinmux control device structure */
+	const struct device *pinctrls;
```
not related to this PR, but be aware that drivers should start moving to `pinctrl`, see https://github.com/zephyrproject-rtos/zephyr/issues/39740

RuibinChang replied:
Ack

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,256 @@
+	uint32_t capture;
+};
+
+static int tach_ch_valid(const struct device *dev, int tach_ch)
```
please use `bool` here

non-blocking suggestion: `tach_ch_is_valid`

RuibinChang replied:
Done

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,256 @@
+	int tach_ch = config->channel;
+	struct tach_it8xxx2_data *const data = dev->data;
+
+	if (chan != SENSOR_CHAN_RPM) {
```
`SENSOR_CHAN_ALL` also needs to be supported in fetch

RuibinChang replied:
Done

gmarull commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -808,6 +809,30 @@
+			reg = <0x00f0181e 1   /* F1TLRR */
+			       0x00f0181f 1   /* F1TMRR */
+			       0x00f01848 1>; /* TSWCTLR */
+			dvs-bit = <0x8>;   /* Bit(3) */
+			chsel-bit = <0x4>; /* Bit(2) */
```
consider using `BIT(n)` from `dt-util.h`

RuibinChang replied:
Done

gmarull commented on dts/riscv/it8xxx2.dtsi:
```diff
@@ -808,6 +809,30 @@
+			chsel-bit = <0x4>; /* Bit(2) */
+			label = "TACH_0";
+			status = "disabled";
+			pinctrl-0 = <&pinctrl_tach0a   /* GPD6 */
+				     &pinctrl_tach0b>; /* GPJ2 */
```
this belongs to board-level DT, will likely need to be addressed when porting to pinctrl.

RuibinChang replied:
Ack

mbolivar-nordic reviewed (approved):
My comments on this PR are addressed.

I am curious about one last thing, though:

```
			reg = <0x00f01820 1   /* F2TLRR */
			       0x00f01821 1   /* F2TMRR */
			       0x00f01848 1>; /* TSWCTLR */
```

Why `1` for the size? That's kind of weird value. Is the register block really 1 byte long?

RuibinChang commented:
> My comments on this PR are addressed.
> 
> I am curious about one last thing, though:
> 
> ```
> 			reg = <0x00f01820 1   /* F2TLRR */
> 			       0x00f01821 1   /* F2TMRR */
> 			       0x00f01848 1>; /* TSWCTLR */
> ```
> 
> Why `1` for the size? That's kind of weird value. Is the register block really 1 byte long?

Our prescaler, PWM and tachometer module register are mess-up in 0xf01800~0xf018ff without any rules, so I list the register one by one for tachometer node.

RuibinChang commented:
About this error: MAINTAINERS.yml: glob pattern 'drivers / * / * it8xxx2 *' in 'files' in area 'ITE Platforms' matches a directory, but has no trailing '/'
Why does the checker always check the original directory 'drivers / * / * it8xxx2 *' instead of added path 'drivers/ * / * / * it8xxx2 *' ?

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,251 @@
+	int dvs_bit = config->dvs_bit;
+	int chsel_bit = config->chsel_bit;
+	int mask = (dvs_bit | chsel_bit);
+	bool valid = 0;
```
`= false;`

RuibinChang replied:
Done

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,251 @@
+		break;
+	case TACH_CHANNEL_B:
+		if ((*reg_tswctlr & mask) == mask) {
+			valid = 1;
```
` = true;`

RuibinChang replied:
Done

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,251 @@
+	volatile uint8_t *reg_fxtmrr = (uint8_t *)config->reg_fxtmrr;
+	volatile uint8_t *reg_tswctlr = (uint8_t *)config->reg_tswctlr;
+	int tach_ch = config->channel;
+	struct tach_it8xxx2_data *const data = dev->data;
+
+	if (tach_ch_is_valid(dev, tach_ch)) {
```
please validate chan here

RuibinChang replied:
Done

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,253 @@
+	int tach_ch = config->channel;
+	struct tach_it8xxx2_data *const data = dev->data;
+
+	if (chan == SENSOR_CHAN_RPM) {
```
you also need to support fetching `SENSOR_CHAN_ALL`, see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L508

RuibinChang replied:
Done

gmarull commented on include/dt-bindings/sensor/it8xxx2_tach.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_SENSOR_ITE_TACH_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_SENSOR_ITE_TACH_H_
+
+/* Tachometer channel references */
+#define TACH_CHANNEL_A		0
+#define TACH_CHANNEL_B		1
```
nit: better namespace these definitions (`IT8XXX2_...`), also add Doxygen docstrings:

```c
/**
 * @name Tachometer channels
 * @{
 */

/** Tachometer channel A */
#define TACH_CHANNEL_A		0
/** Tachometer channel B */
#define TACH_CHANNEL_B		1

/** @} */
```

RuibinChang replied:
Done

gmarull commented on soc/riscv/riscv-ite/common/chip_chipregs.h:
```diff
@@ -44,6 +44,9 @@
 #define REG_BASE_ADDR				EC_REG_BASE_ADDR
 #endif
 
+/* Common definition */
+#define EC_FREQ			MHZ(8)
```
can the purpose/meaning be explained better here?

RuibinChang replied:
Done

gmarull commented on drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c:
```diff
@@ -0,0 +1,253 @@
+	int tach_ch = config->channel;
+	struct tach_it8xxx2_data *const data = dev->data;
+
+	if ((chan == SENSOR_CHAN_RPM) || (chan == SENSOR_CHAN_ALL)) {
```
nit:

if wrong channel is provided, this function will still return 0, should return `-ENOTSUP`.

you can do:
```c
if ((chan != SENSOR_CHAN_RPM) && (chan != SENSOR_CHAN_ALL)) {
    return -ENOTSUP;
}
```

RuibinChang replied:
Done

mbolivar-nordic reviewed (approved):
For the binding

RuibinChang commented:
Could we merge this PR? The PR https://github.com/zephyrproject-rtos/zephyr/pull/40274 needs the EC_FREQ common definition. Or need I do the changes for Compliance Checks / Check MAINTAINERS file?

jackrosenthal commented:
> The PR #40274 needs the EC_FREQ common definition.

It seems like that PR depended on this one, but they were not uploaded together. Please revert that PR until this one lands.

RuibinChang commented:
Ack

nashif commented on MAINTAINERS.yml:
```diff
@@ -1570,6 +1570,7 @@ ITE Platforms:
         - sjg20
     files:
         - boards/riscv/it8xxx2_evb/
+        - drivers/*/*/*it8xxx2*
```
what are you trying to match here? if `drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c`, then you should adapt the other wildcard to match the files explicitly, otherwise you will end up with a conflict.

RuibinChang replied:
Yes, I want to match the drivers/.../tach_ite_it8xxx2.c path.
Done, thanks.
